fw/http.c

```diff
@@ -747,27 +747,35 @@ tfw_http_prep_redir(TfwHttpResp *resp, unsigned short status, TfwStr *rmark,
 	if (host.len) {
 		url.chunks[url.nchunks++] = *proto;
 		url.len += proto->len;
-
-		url.chunks[url.nchunks++] = host;
-		url.len += host.len;
+		host_val = *this_cpu_ptr(&g_buf) + SLEN(S_V_DATE) + cl_len;
+		host_len = tfw_str_to_cstr(&host, host_val, RESP_BUF_LEN - SLEN(S_V_DATE) - cl_len);
+		url.chunks[url.nchunks].data = host_val;
+		url.chunks[url.nchunks].len = host_len;
+		url.len += host_len;
+		url.nchunks++;
 	}
 
 	if (rmark->len) {
 		url.chunks[url.nchunks++] = *rmark;
 		url.len += rmark->len;
 	}
 
-	url.chunks[url.nchunks++] = req->uri_path;
-	url.len += req->uri_path.len;
+	uri_path_val = *this_cpu_ptr(&g_buf) + SLEN(S_V_DATE) + cl_len + host_len;
+	uri_path_len = tfw_str_to_cstr(&req->uri_path, uri_path_val, RESP_BUF_LEN - SLEN(S_V_DATE) - cl_len - host_len);
+	url.chunks[url.nchunks].data = uri_path_val;
+	url.chunks[url.nchunks].len = uri_path_len;
+	url.len += uri_path_len;
+	url.nchunks++;
 
 	/*
 	 * We have to copy the body since tfw_h2_append_predefined_body() called
 	 * from tfw_h2_frame_local_resp() expects body as plain string.
 	 * At the moment this function is used for sticky session redirects, so
 	 * there is no big difference wehre to copy the body.
 	 */
+	body_val = *this_cpu_ptr(&g_buf) + SLEN(S_V_DATE) + cl_len + host_len + uri_path_len;
```

Comment:
Why not to get a pointer and move it forward after each write to `g_buf` instead of repetition of the shift calculation?

Also in the origin code we had only date value and the body size, which are definitely smaller than 2KB, but this is doubtful with the Host value, so in this patch we have to track whether we fit into `g_buf`.

Reply:
fixed